File: frontend/lib/features/books/data/books_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../core/network/providers.dart';
import '../../../data/models/book.dart';
import '../../../data/models/external_book_result.dart';
import '../../../data/models/map_city.dart';
import '../../../data/models/price_offer.dart';
import '../../../data/models/user_book.dart';

class BrowseResult {
  final List<UserBook> items;
  final int total;

  const BrowseResult({required this.items, required this.total});
}

class ViewStats {
  final int total;
  final int unique;

  const ViewStats({required this.total, required this.unique});

  factory ViewStats.fromJson(Map<String, dynamic> json) {
    return ViewStats(total: json['total'] as int, unique: json['unique'] as int);
  }
}

class BooksRepository {
  BooksRepository(this._ref);
  final Ref _ref;

  Future<BrowseResult> browse({
    String? title,
    String? author,
    String? genre,
    String? language,
    String? city,
    String? condition,
    String? sort,
    String? fromCity,
    int? maxDistanceKm,
    int limit = 20,
    int offset = 0,
  }) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/browse', queryParameters: {
      if (title != null && title.isNotEmpty) 'title': title,
      if (author != null && author.isNotEmpty) 'author': author,
      if (genre != null && genre.isNotEmpty) 'genre': genre,
      if (language != null && language.isNotEmpty) 'language': language,
      if (city != null && city.isNotEmpty) 'city': city,
      if (condition != null && condition.isNotEmpty) 'condition': condition,
      if (sort != null && sort.isNotEmpty) 'sort': sort,
      if (fromCity != null && fromCity.isNotEmpty) 'fromCity': fromCity,
      'maxDistanceKm': ?maxDistanceKm,
      'limit': limit,
      'offset': offset,
    });

    final items = (response.data['items'] as List<dynamic>)
        .map((e) => UserBook.fromJson(e as Map<String, dynamic>))
        .toList();

    return BrowseResult(items: items, total: response.data['total'] as int);
  }

  Future<UserBook> getUserBook(String id) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/$id');
    return UserBook.fromJson(response.data as Map<String, dynamic>);
  }

  Future<ViewStats> getViewStats(String id) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/$id/views');
    return ViewStats.fromJson(response.data as Map<String, dynamic>);
  }

  Future<List<ListingHistoryEntry>> getListingHistory(String id) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/$id/history');
    return (response.data as List)
        .map((e) => ListingHistoryEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<UserBook>> getSimilarBooks(String id) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/$id/similar');
    return (response.data as List)
        .map((e) => UserBook.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<BookGenre>> getGenres({String? query}) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/genres', queryParameters: {
      if (query != null && query.isNotEmpty) 'query': query,
    });
    return (response.data as List)
        .map((e) => BookGenre.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<String>> getAuthorSuggestions(String query) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/authors', queryParameters: {
      if (query.isNotEmpty) 'query': query,
    });
    return (response.data as List).cast<String>();
  }

  Future<List<String>> getLanguageSuggestions(String query) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/languages', queryParameters: {
      if (query.isNotEmpty) 'query': query,
    });
    return (response.data as List).cast<String>();
  }

  Future<List<BookStatEntry>> getMostSharedBooks() async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/most-shared');
    return (response.data as List)
        .map((e) => BookStatEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<BookStatEntry>> getTrendingBooks() async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/trending');
    return (response.data as List)
        .map((e) => BookStatEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<AuthorStatEntry>> getMostPopularAuthors() async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/popular-authors');
    return (response.data as List)
        .map((e) => AuthorStatEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<SearchStat>> getPopularSearches() async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/popular-searches');
    return (response.data as List)
        .map((e) => SearchStat.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<UserBook>> getNearbyToday(String city) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/nearby-today', queryParameters: {'city': city});
    return (response.data as List)
        .map((e) => UserBook.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<MapCity>> getMapCities() async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/map-cities');
    return (response.data as List)
        .map((e) => MapCity.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<UserBook>> getMyLibrary() async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/my-library');
    return (response.data as List<dynamic>)
        .map((e) => UserBook.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<UserBook> setAvailability(String userBookId, {required bool availableForSwap}) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.patch(
      '/books/$userBookId',
      data: {'availableForSwap': availableForSwap},
    );
    return UserBook.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteUserBook(String userBookId) async {
    final dio = _ref.read(apiClientProvider).dio;
    await dio.delete('/books/$userBookId');
  }

  Future<List<ExternalBookResult>> searchExternal(String query) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.get('/books/search', queryParameters: {'q': query});
    return (response.data as List<dynamic>)
        .map((e) => ExternalBookResult.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<UserBook> addToLibrary({
    String? isbn,
    String? title,
    String? author,
    required BookCondition condition,
    String? language,
    String? edition,
    bool isHardcover = false,
  }) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.post('/books', data: {
      if (isbn != null && isbn.isNotEmpty) 'isbn': isbn,
      if (title != null && title.isNotEmpty) 'title': title,
      if (author != null && author.isNotEmpty) 'author': author,
      'condition': condition.toJson(),
      if (language != null && language.isNotEmpty) 'language': language,
      if (edition != null && edition.isNotEmpty) 'edition': edition,
      'isHardcover': isHardcover,
    });
    return UserBook.fromJson(response.data as Map<String, dynamic>);
  }

  Future<UserBook> relistBook(
    String originalUserBookId, {
    required BookCondition condition,
    String? language,
    String? edition,
    bool isHardcover = false,
  }) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.post('/books/$originalUserBookId/relist', data: {
      'condition': condition.toJson(),
      if (language != null && language.isNotEmpty) 'language': language,
      if (edition != null && edition.isNotEmpty) 'edition': edition,
      'isHardcover': isHardcover,
    });
    return UserBook.fromJson(response.data as Map<String, dynamic>);
  }

  /// Trece un anunț la vânzare - separat de creare, pentru că backend-ul
  /// cere cel puțin o poză deja urcată (vezi updateUserBook), iar la
  /// creare nu poate exista încă nicio poză.
  Future<UserBook> markForSale(
    String userBookId, {
    required double salePrice,
    required bool isNegotiable,
  }) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.patch('/books/$userBookId', data: {
      'isForSale': true,
      'salePrice': salePrice,
      'isNegotiable': isNegotiable,
    });
    return UserBook.fromJson(response.data as Map<String, dynamic>);
  }

  /// Editare completă a unui anunț deja publicat - condiție, ediție, limbă,
  /// copertă cartonată și, dacă e la vânzare, preț/negociabil. Folosește
  /// același PATCH ca markForSale/setAvailability (backend-ul validează
  /// oricum că există cel puțin o poză înainte de a permite isForSale=true).
  Future<UserBook> updateListing(
    String userBookId, {
    required BookCondition condition,
    String? language,
    String? edition,
    required bool isHardcover,
    required bool isForSale,
    double? salePrice,
    required bool isNegotiable,
  }) async {
    final dio = _ref.read(apiClientProvider).dio;
    final response = await dio.patch('/books/$userBookId', data: {
      'condition': condition.toJson(),
      'language': language,
      'edition': edition,
      'isHardcover': isHardcover,
      'isForSale': isForSale,
      if (isForSale) 'salePrice': salePrice,
      'isNegotiable': isNegotiable,
    });
    return UserBook.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> addPhoto(String userBookId, {required List<int> bytes, required String filename}) async {
    final dio = _ref.read(apiClientProvider).dio;
    final formData = FormData.fromMap({
      'photo': MultipartFile.fromBytes(bytes, filename: filename.isEmpty ? 'photo.jpg' : filename),
    });
    await dio.post('/books/$userBookId/photos', data: formData);
  }
}

final booksRepositoryProvider = Provider<BooksRepository>((ref) {
  return BooksRepository(ref);
});

File: web/src/features/books/booksRepository.ts
import { api } from '@/lib/api/client';
import type { UserBook } from '@/types/models';

export interface BrowseResult {
  items: UserBook[];
  total: number;
}

export interface BrowseParams {
  title?: string;
  author?: string;
  genre?: string;
  language?: string;
  city?: string;
  condition?: string;
  sort?: string;
  fromCity?: string;
  maxDistanceKm?: number;
  excludeUserId?: string;
  listingType?: string;
  limit?: number;
  offset?: number;
}

export const booksRepository = {
  browse(params: BrowseParams = {}, signal?: AbortSignal): Promise<BrowseResult> {
    const { limit = 20, offset = 0, ...filters } = params;
    // Parametrii goi nu se trimit deloc: backendul tratează `genre=""` drept
    // filtru pe genul gol, nu drept "fără filtru", și întoarce zero rezultate.
    const query: Record<string, string | number> = { limit, offset };
    for (const [key, value] of Object.entries(filters)) {
      if (value !== undefined && value !== null && value !== '') query[key] = value;
    }
    return api.get<BrowseResult>('/books/browse', { query, signal });
  },

  getById(userBookId: string, signal?: AbortSignal): Promise<UserBook> {
    return api.get<UserBook>(`/books/${userBookId}`, { signal });
  },

  getRecommended(signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>('/books/recommended', { signal });
  },

  getSimilar(userBookId: string, signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>(`/books/${userBookId}/similar`, { signal });
  },

  // --- secțiunile din Descoperă ---
  // Fiecare e un GET separat, invalidabil individual, exact ca providerii din
  // discover_screen.dart. Nu le unim într-un singur endpoint: o secțiune lentă
  // (recomandările, care fac muncă reală pe backend) ar ține în loc tot ecranul.

  getTrendingListings(signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>('/books/trending-listings', { signal });
  },

  getMostWished(signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>('/books/most-wished', { signal });
  },

  getHiddenGems(signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>('/books/hidden-gems', { signal });
  },

  getNearbyToday(city: string, signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>('/books/nearby-today', { query: { city }, signal });
  },

  getPopularSearches(signal?: AbortSignal): Promise<SearchStat[]> {
    return api.get<SearchStat[]>('/books/popular-searches', { signal });
  },

  getPopularAuthors(signal?: AbortSignal): Promise<AuthorStat[]> {
    return api.get<AuthorStat[]>('/books/popular-authors', { signal });
  },

  getGenres(signal?: AbortSignal): Promise<GenreStat[]> {
    return api.get<GenreStat[]>('/books/genres', { signal });
  },

  getMyLibrary(signal?: AbortSignal): Promise<UserBook[]> {
    return api.get<UserBook[]>('/books/my-library', { signal });
  },

  /** Caută în catalog + surse externe, pentru precompletarea formularului. */
  search(query: string, signal?: AbortSignal): Promise<ExternalBookResult[]> {
    return api.get<ExternalBookResult[]>('/books/search', { query: { q: query }, signal });
  },

  lookupIsbn(isbn: string, signal?: AbortSignal): Promise<ExternalBookResult | null> {
    return api.get<ExternalBookResult | null>('/books/lookup-isbn', {
      query: { isbn },
      signal,
    });
  },

  suggestCovers(input: { title: string; author?: string }, signal?: AbortSignal) {
    return api.get<string[]>('/books/covers', {
      query: { title: input.title, ...(input.author ? { author: input.author } : {}) },
      signal,
    });
  },

  /**
   * Câmpurile goale NU se trimit deloc. Backendul le-ar salva literal ca "",
   * iar cartea ar apărea cu o editură goală în loc să n-o afișeze.
   */
  addToLibrary(input: AddBookInput): Promise<UserBook> {
    const payload: Record<string, unknown> = { isHardcover: input.isHardcover ?? false };
    for (const [key, value] of Object.entries(input)) {
      if (key === 'isHardcover') continue;
      if (value === undefined || value === null || value === '') continue;
      if (Array.isArray(value) && value.length === 0) continue;
      payload[key] = value;
    }
    return api.post<UserBook>('/books', payload);
  },

  update(userBookId: string, input: Record<string, unknown>): Promise<UserBook> {
    return api.patch<UserBook>(`/books/${userBookId}`, input);
  },

  markForSale(userBookId: string, salePrice: number, isNegotiable: boolean): Promise<UserBook> {
    return api.patch<UserBook>(`/books/${userBookId}`, {
      isForSale: true,
      salePrice,
      isNegotiable,
    });
  },

  async addPhoto(userBookId: string, file: File): Promise<{ photoUrl?: string }> {
    const form = new FormData();
    form.append('photo', file);
    return api.request<{ photoUrl?: string }>(`/books/${userBookId}/photos`, {
      method: 'POST',
      formData: form,
    });
  },

  addPhotoFromUrl(userBookId: string, url: string): Promise<void> {
    return api.post(`/books/${userBookId}/photos/from-url`, { url });
  },

  /**
   * Poza principală a anunțului - cea care apare în feed. URL-ul poate fi o
   * poză urcată sau o copertă externă; `null` revine la fallback.
   */
  setMainPhoto(userBookId: string, mainPhotoUrl: string | null): Promise<void> {
    return api.patch(`/books/${userBookId}`, { mainPhotoUrl });
  },

  /**
   * Scorurile de anunt, pentru badge-ul vizibil DOAR adminilor.
   *
   * E un POST cu o lista, nu un GET per card: o grila de discover are 20-30 de
   * carduri, iar cate o cerere de fiecare ar insemna 30 de dus-intors la
   * fiecare derulare. Backendul limiteaza la 100 de id-uri per cerere.
   */
  getListingScores(userBookIds: string[]): Promise<Record<string, number>> {
    if (userBookIds.length === 0) return Promise.resolve({});
    return api.post<Record<string, number>>('/books/scores', { userBookIds });
  },
};

/** Rezultat de căutare, din catalog sau dintr-o sursă externă. */
export interface ExternalBookResult {
  id?: string;
  isbn?: string | null;
  title: string;
  author?: string | null;
  coverUrl?: string | null;
  publisher?: string | null;
  publishedYear?: number | null;
  pageCount?: number | null;
  description?: string | null;
  genre?: string | null;
  language?: string | null;
}

export interface AddBookInput {
  bookId?: string;
  isbn?: string;
  title?: string;
  author?: string;
  language?: string;
  edition?: string;
  isHardcover?: boolean;
  condition?: string;
  genre?: string;
  series?: string;
  seriesNumber?: number;
  publisher?: string;
  publishedYear?: number;
  editionYear?: number;
  pageCount?: number;
  description?: string;
  tags?: string[];
  city?: string;
  mainPhotoUrl?: string;
}

export interface SearchStat {
  query: string;
  count: number;
}

export interface AuthorStat {
  author: string;
  count: number;
}

export interface GenreStat {
  genre: string;
  count: number;
}

/**
 * Cheile de cache. Grupate într-un singur loc ca invalidarea după o mutație să
 * nu fie o ghicitoare - un string scris ușor diferit în două fișiere înseamnă
 * două intrări separate în cache, iar ecranul rămâne pe datele vechi fără
 * nicio eroare vizibilă.
 */
export const booksKeys = {
  all: ['books'] as const,
  browse: (params: BrowseParams) => ['books', 'browse', params] as const,
  detail: (id: string) => ['books', 'detail', id] as const,
  recommended: () => ['books', 'recommended'] as const,
  similar: (id: string) => ['books', 'similar', id] as const,
  trending: () => ['books', 'trending-listings'] as const,
  mostWished: () => ['books', 'most-wished'] as const,
  hiddenGems: () => ['books', 'hidden-gems'] as const,
  nearby: (city: string) => ['books', 'nearby-today', city] as const,
  popularSearches: () => ['books', 'popular-searches'] as const,
  popularAuthors: () => ['books', 'popular-authors'] as const,
  genres: () => ['books', 'genres'] as const,
  myLibrary: () => ['books', 'my-library'] as const,
};
